packages/core/lib/index.js

```diff
@@ -124,28 +124,34 @@ const evalRule = async (rule, scope) => {
     // async This try...catch will handle that, and also any errors from a
     // rejected promise
     const promise = Promise.resolve(
-      resolveFn(fn.name, scope)(scope)(resolvedOpts)(...resolvedArgs)
+      resolveRule(fn.name, scope)(scope)(resolvedOpts)(...resolvedArgs)
     );
     return await promise;
   } catch (err) {
     throw BrowseError.from(err, fn.name);
   }
 };
 
-const evalRuleSet = async (ruleSet, parent) => {
-  assert(ruleSet.type === "RuleSet");
-
-  if (!parent) {
-    parent = getSTD({ evalRule, evalRuleSet, getNewScope });
-  }
-
-  const { rules: oRules } = ruleSet;
-  const rules = [...oRules]; // Don't want to modify the original rules
+const evalRuleSet = async (ruleSet, inject = {}) => {
+  const rules = [...ruleSet.rules]; // Don't want to modify the original rules
 
-  const scope = getNewScope(parent);
   if (!rules.length) {
     return null;
   }
+
+  // Setup scope
+  let scope;
+  if (inject.parent !== undefined) {
+    // if inject has a `parent` set, then we just treat inject as the scope
```

Comment:
For now I would console.warn a warning "Implicit scope injection used". Otherwise if someone hits this logic when they shouldn't it's going to be really hard to debug.

Reply:
no we actually do already use it. For example, I use this when visit creates and injects a page scope into the page ruleSet, and the `for` rule has to use this too

It's just not the recommended way (which is why I said rarely), because if someone doesn't use this, then it's upto the caller to correctly implement lexical scoping. That's why you'll see in both those places where I used this to control the scope, I specifically managed the inheritance myself so that the scope that was used here inherited from the correct lexical scope

the comment here is to discourage you from using it this way, so that `evalRuleset` can handle the lexical scoping